r/src/arrow_types.h

```diff
@@ -176,64 +37,21 @@ inline constexpr Rbyte default_value<RAWSXP>() {
 
 }  // namespace Rcpp
 
-namespace arrow {
-namespace r {
-
-template <typename T>
-inline std::shared_ptr<T> extract(SEXP x) {
-  return Rcpp::ConstReferenceSmartPtrInputParameter<std::shared_ptr<T>>(x);
-}
-
-}  // namespace r
-}  // namespace arrow
-
 #if defined(ARROW_R_WITH_ARROW)
-#include <arrow/api.h>
-#include <arrow/c/bridge.h>
-#include <arrow/compute/api.h>
-#include <arrow/csv/reader.h>
-#include <arrow/dataset/api.h>
-#include <arrow/filesystem/filesystem.h>
-#include <arrow/filesystem/localfs.h>
-#include <arrow/filesystem/s3fs.h>
-#include <arrow/io/compressed.h>
-#include <arrow/io/file.h>
-#include <arrow/io/memory.h>
-#include <arrow/ipc/feather.h>
-#include <arrow/ipc/reader.h>
-#include <arrow/ipc/writer.h>
-#include <arrow/json/reader.h>
+#include <arrow/buffer.h>  // for RBuffer definition below
```

Comment:
The RBuffer class definition should probably be moved to buffer.cpp but I didn't work out how to register it in the namespace in this file.

Reply:
`RBuffer` is a templated class, the definition must be in the header and all it's dependencies must also be included in the headers.

In other words, by moving `RBuffer` in it's own file, the using classes will also need to include this new header + the original `<arrow/buffer.h>`, e.g. no benefits.

Reply:
Right but only 2 cpp files need RBuffer, so we wouldn’t have to include it for the other 20 files. So does that help?

Reply:
It doesn't, I had to touch 6 files, but it's exactly the same time to compile.

Reply:
👍 I don't have great instincts about what drives compile time and how to optimize it